addons/delivery/models/delivery_carrier.py

```diff
@@ -277,3 +278,111 @@ def fixed_get_tracking_link(self, picking):
+            unit_quantity = line.product_uom._compute_quantity(line.product_uom_qty, line.product_id.uom_id, rounding_method='HALF-UP')
+            rounded_qty = max(1, float_round(unit_quantity, precision_digits=0, rounding_method='HALF-UP'))
+            country_of_origin = line.product_id.country_of_origin or order.warehouse_id.partner_id.country_id.code
+            commodities.append(DeliveryCommodity(line.product_id, amount=rounded_qty, monetary_value=line.price_reduce_taxinc, country_of_origin=country_of_origin))
+
+        return commodities
+
+    def _get_commodities_from_stock_move_lines(self, move_lines):
+        commodities = []
+
+        for line in move_lines.filtered(lambda line: line.product_id.type in ['product', 'consu']):
+            if line.state == 'done':
+                unit_quantity = line.product_uom_id._compute_quantity(line.qty_done, line.product_id.uom_id, rounding_method='HALF-UP')
+            else:
+                unit_quantity = line.product_uom_id._compute_quantity(line.product_uom_qty, line.product_id.uom_id, rounding_method='HALF-UP')
+            rounded_qty = max(1, float_round(unit_quantity, precision_digits=0, rounding_method='HALF-UP'))
+            country_of_origin = line.product_id.country_of_origin or line.picking_id.picking_type_id.warehouse_id.partner_id.country_id.code
+            commodities.append(DeliveryCommodity(line.product_id, amount=rounded_qty, monetary_value=line.sale_price, country_of_origin=country_of_origin))
+
+        return commodities
+
+    def _product_price_to_company_currency(self, quantity, product, company):
+        return company.currency_id._convert(quantity * product.standard_price, product.currency_id, company, fields.Date.today())
+
+
+class DeliveryPackage:
+    """ Each provider need similar information about its packages. """
+    def __init__(self, commodities, weight, package_type, name=None, total_cost=0, currency=None):
+        """ The UOMs are based on the config parameters, which is very convenient:
+        we do not need to keep those stored."""
+        self.commodities = commodities or []  # list of DeliveryCommodity objects
+        self.weight = weight
+        self.dimension = {
+            'length': package_type.packaging_length,
+            'width': package_type.width,
+            'height': package_type.height
+        }
+        self.packaging_type = package_type.shipper_package_code or False
+        self.name = name
+        self.total_cost = total_cost
+        self.currency_id = currency
+
+
+class DeliveryCommodity:
+    """ Commodities information are needed for Commercial invoices with each provider. """
+    def __init__(self, product, amount, monetary_value, country_of_origin):
+        self.product_id = product
+        self.qty = amount
+        self.monetary_value = monetary_value  # based on company currency
+        self.country_of_origin = country_of_origin
```

Comment:
Put them in another file

Reply:
Something like `delivery_customs.py` ?

Reply:
I would create a delivery_request_package and a delivery_request_commodity or delivery_request_objects for both